Ds 498

Fennersteel opened the pull request.

Fennersteel committed:
First finished version of irods_move2 python script. Needs addition of file rename function

Fennersteel committed:
Version of test_irods_move.yml now including irods test env info:
	- Added host, port, admin_user, admin_password, and zone arguments to each task.

Fennersteel committed:
Final version of python implementation of irods_move script
	- Added documentation for EXAMPLES and RETURNS
	- Added functionality to return exception type and error message when errors occur

Fennersteel committed:
ds-498: Final version of test_irods_move.yml

Fennersteel committed:
ds-498: Removed undesired changes to files in irods library

tedgin commented on irods/library/tests/test_irods_move.yml:
```diff
@@ -2,6 +2,101 @@
+        IRODS_PASSWORD: rods
+        IRODS_ZONE_NAME: testing
+
+    # - name: source and destination already exist
```
Don't commit commented out code. It clutters up the source.

tedgin commented on irods/library/tests/test_irods_move.yml:
```diff
@@ -2,6 +2,101 @@
+      register: response
+
+  post_tasks: 
+    - name: verify move 
```
After some consideration, the intent of this task would be more clear if it were called immediately after the `move data object` task.

tedgin commented on irods/library/tests/test_irods_move.yml:
```diff
@@ -2,6 +2,101 @@
-    - debug:
-        msg: TODO implement
\ No newline at end of file
+  vars:
```
Since you've declared these constants, they should be used in the subsequent tasks instead of their literal values.

tedgin commented on irods/library/tests/test_irods_move.yml:
```diff
@@ -2,6 +2,101 @@
+          set -o errexit
+          iinit <<< ${IRODS_PASSWORD}
+          imkdir {{ TEST_SRC }}
+          ils
```
Remove `ils`. We put it in for debugging purposes.

tedgin commented on irods/library/tests/test_irods_move.yml:
```diff
@@ -2,6 +2,101 @@
+        zone: testing
+      register: response
+
+    - name: create test collection
```
This task is setting up for the testing. It's not actually a test.  It should be merged into the `create source` pretask.

tedgin commented on irods/library/tests/test_irods_move.yml:
```diff
@@ -2,6 +2,101 @@
+          iinit <<< ${IRODS_PASSWORD}
+          imkdir {{ TEST_SRC }}
+          ils
+      environment:
```
These environment variables are used in several tasks, so move their declaration to the play's definition to reduce the size of the code.

tedgin replied:
I meant that you should remove the `environment` fields attached to the tasks add one to the play. For example,

```
- name: test irods_move module
  ...
  vars:
    ...
  environment:
    IRODS_PORT: 1247
    ...
  pre_tasks:
    - name: create test objects
      shell: ...
      tags:
        - non_idempotent
  ...
```

tedgin commented on irods/library/tests/test_irods_move.yml:
```diff
@@ -2,6 +2,101 @@
+    #     zone: testing
+    #   register: response
+
+    - name: move data object
```
This task is moving a collection, not a data object. Please rename the task.

tedgin committed:
DS-498: adding irods/library path for ansible-lint

tedgin committed:
DS-498: adapting current playbooks to new irods_move

tedgin reviewed (changes requested):
I've only reviewed the ansible changes.

Fennersteel committed:
Made requisite changes after initial pull request
	- Removed committed out code at line 48.
	- Renamed "Move data object" task to "Move data collection".
	- Moved the "verify move" task to directly after the "move data collection" task.
	- Refactored code to make use of defined variables in "vars" rather than using literal values.
	- Removed unnecessary code used for debugging.

Fennersteel committed:
Merge branch 'ds-498' of https://github.com/cyverse/ds-playbooks into ds-498

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,175 @@
+
+short_description: Rename data collection/object
+
+# If this is part of a collection, you need to use semantic versioning,
```
Don't put Python comments in the module's documentation.

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,175 @@
+
+# If this is part of a collection, you need to use semantic versioning,
+# i.e. the version is of the form "2.5.0" and not "2.4".
+version_added: "1.0.0"
```
This is the version of Ansible you were using when you created this module.

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,175 @@
+    returned: always
+'''
+
+try:
```
Since the python-irodsclient modules are imported outside of this try-except clause, it's not needed.  `USE_IRODS_CLIENT` will always be set to `True`.  Either move the python-irodsclient module import inside the try block, or remove the try-except clause and usage of the variable `USE_IRODS_CLIENT`.

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,175 @@
+try:
+    env_file = os.environ['IRODS_ENVIRONMENT_FILE']
+except KeyError:
+    env_file = os.path.expanduser('~/.irods/irods_environment.json')
```
The variable `env_file` isn't used please remove this try-except clause.

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,175 @@
+except KeyError:
+    env_file = os.path.expanduser('~/.irods/irods_environment.json')
+
+ssl_context = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH,
```
This variable and the following aren't used. Please remove them.

tedgin reviewed (changes requested):
In addition to the specific issues mentioned inline, please address the ansible-lint, pylint, and pycodestyle issues.

Fennersteel committed:
- test_irods_move.yml
	- Added environment classification to the play

- irods_move
	- Removed unnecessary comments
	- Rectified documentation to display correct ansible version used in development
	- Removed unused variables

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,157 @@
+
+        self.ssl_settings = {"ssl_context": self.ssl_context}
+
+        self.session = iRODSSession(
```
If this throws an exception, the module will exit without notifying Ansible. This could happen if the function can't connect to iRODS. You should wrap this in  a try-except, where the except block calls self.module.fail_json and rethrows the exception.

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,157 @@
+    """
+    Entrypoint of the Ansible module
+    """
+    module = IRODSMoveModule()
```
The contents of main should be wrapped in a try-except block to prevent the exception from escaping the module.  The except should just call `pass`.

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,157 @@
+
+        try:
+            self.rename()
+            self.result["changed"] = True
```
Remove this line. The `changed` result should be set in `rename()`.

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,157 @@
+        # rename collection
+        path = self.module.params["source"]
+        new_path = self.module.params["destination"]
+        self.session.collections.move(path, new_path)
```
This function needs to verify that `path` exists and `new_path` does not before calling `session.collections.move`. If `path` does not exist, and  `new_path` does, then it should assume the move has already happened and the `changed` result left as `False`.

Fennersteel committed:
Added checks against common errors and move script running twice

- irods_move:
	- Added several try/except blocks to common places of error to ensure the script exits properly
	- Removed redundant lines of code
	- Added functionality to prevent script from trying to run when the given source path doesn't exist and the given destination path does

tedgin commented on irods/library/irods_move:
```diff
@@ -1,142 +1,177 @@
+        path_exists = self.session.collections.exists(path)
+        new_path_exists = self.session.collections.exists(new_path)
+
+        if path_exists is True and new_path_exists is False:
```
There are four cases:

1) source and destination exist  -- This is an error, won't overwrite existing collection or data object
2) source exists but destination doesn't -- This is when collections.move is called and change = True
3) source doesn't exist but destination does -- This is not an error, we silently assume the move already happened, change = False; we make this assumption to support idempotency.
4) neither source nor destination exist. -- This is an error, source doesn't exist

You only need to test for 3).  collections.move will handle cases 1), 2), and 4) correctly.

Fennersteel committed:
Cleaned up the logic in irods_move

- irods_move
	- Cleaned up script logic by removing unnecessary if/else clauses in rename function
- test_irods_move.yml
	- Deleted unused test task

tedgin committed:
DS-498: linting changes for test-playbook.sh

tedgin committed:
DS-498: testing-playbook.sh now fails idempotency when task fails

tedgin commented on irods/library/tests/test_irods_move.yml:
```diff
@@ -2,6 +2,66 @@
+        zone: testing
+      register: response
+
+    - name: verify move 
```
This task doesn't change anything.  Ansible doesn't know whether a `shell` task makes a change, so it assumes it does.  To tell ansible that no changes were made, at the `changed_when: false` option to the task description.

Fennersteel committed:
Updated test_irods_move verification task
- test_irods_move.yml
	- Added a changed_when condition to the "verify_move" task to help ansible verify when a change occurs in the file system.
